File: packages/ee-mission/test/spawn-diagnostics.lua
-- Engine-boundary fault injection. Does not simulate DCS parking or prove takeoff.
local root = arg[1] or '.'
local file = assert(io.open(root .. '/dist/ee-dcs.lua', 'rb'))
local bundle = file:read('*a'); file:close()
local replaced
bundle, replaced = bundle:gsub('local ____entry = require%("index", %.%.%.%)%s+return ____entry', 'return require')
assert(replaced == 1)

local function runtime(outcome, debug_enabled)
    local e = setmetatable({}, {__index=_G}); e._G=e; e._DMT_DEBUG=debug_enabled
    local now, logs, timers, handlers, units, groups = 0, {}, {}, {}, {}, {}
    local calls = 0
    local function log(message) logs[#logs+1]=message end
    e.env={info=log}
    e.coalition={side={BLUE=2,RED=1,NEUTRAL=0}}
    e.Group={Category={AIRPLANE=0,HELICOPTER=1,GROUND=2},getByName=function(name) return groups[name] end}
    e.Unit={getByName=function(name) return units[name] end}
    e.StaticObject={getByName=function() return nil end}
    e.Airbase={Category={HELIPAD=1},getByName=function() return nil end}
    e.land={getHeight=function() return 20 end}
    e.world={event={S_EVENT_BIRTH=15,S_EVENT_TAKEOFF=3,S_EVENT_DEAD=8,S_EVENT_CRASH=5}}
    e.world.addEventHandler=function(h) handlers[h]=true end
    e.world.removeEventHandler=function(h) handlers[h]=nil end
    e.timer={getTime=function() return now end,scheduleFunction=function(fn,arg,time) timers[#timers+1]={fn=fn,arg=arg,time=time} end}
    local function event(id, unit)
        for h in pairs(handlers) do h:onEvent({id=id,initiator=unit}) end
    end
    local function materialise(name)
        local unit={alive=true,active=true,air=false}
        function unit:getName() return name end
        function unit:isExist() return self.alive end
        function unit:isActive() return self.active end
        function unit:getLife() return self.alive and 100 or 0 end
        function unit:inAir() return self.air end
        function unit:getPoint() return {x=100,y=25,z=200} end
        function unit:getDesc() return {attributes={Helicopters=true,['Attack helicopters']=true}} end
        units[name]=unit
        event(15,unit)
        return unit
    end
    e.coalition.addGroup=function(_, _, data)
        calls=calls+1
        if outcome=='threw' then error('injected addGroup failure') end
        if outcome=='nil' then return nil end
        local group={}
        function group:isExist() return true end
        function group:getSize()
            local count=0
            for _,u in ipairs(data.units) do if units[u.name] and units[u.name].alive then count=count+1 end end
            return count
        end
        groups[data.name]=group
        if outcome=='full' or outcome=='partial' then materialise(data.units[1].name) end
        if outcome=='full' then materialise(data.units[2].name) end
        return group
    end
    e.coalition.addStaticObject=function() return nil end
    local require_module=setfenv(assert(loadstring(bundle)),e)()
    local queue=require_module('spawn_queue')
    local state=require_module('campaign_state').S
    local function advance(t)
        while true do
            local best
            for i,v in ipairs(timers) do if v.time and v.time<=t and (not best or v.time<timers[best].time) then best=i end end
            if not best then break end
            local v=timers[best]; now=v.time; v.time=v.fn(v.arg,now)
        end
        now=t
    end
    local function contains(text)
        for _,line in ipairs(logs) do if line:find(text,1,true) then return true end end
        return false
    end
    local function spawn()
        e.coalition.addGroup(81,1,{name='test',units={{name='test-1',type='Mi-24V',x=100,y=200},{name='test-2',type='Mi-24V',x=100,y=200}},route={points={{x=100,y=200,type='TakeOffParkingHot',action='From Parking Area Hot',helipadId=77,linkUnit=77}}}})
        assert(queue.drain(log)==1)
    end
    return {e=e,state=state,queue=queue,logs=logs,timers=timers,handlers=handlers,units=units,spawn=spawn,advance=advance,contains=contains,event=event,materialise=materialise,calls=function() return calls end,require_module=require_module}
end

for _,outcome in ipairs({'threw','nil','empty','partial','full'}) do
    local r=runtime(outcome,true)
    r.spawn()
    assert(r.contains('phase=submitted') and r.contains('type=TakeOffParkingHot'))
    assert(r.contains('outcome='..(outcome=='threw' and 'threw' or outcome=='nil' and 'nil' or 'returned')))
    if outcome=='partial' then assert(r.contains('group_size=1') and r.contains('unit=test-2 present=false')) end
    if outcome=='empty' then assert(r.contains('group_size=0') and r.contains('unit=test-1 present=false')) end
    if outcome=='full' then
        assert(r.contains('phase=event id=15'),'birth during addGroup was missed')
        assert(r.contains('group_size=2') and r.contains('agl=5'))
        assert(r.contains('active=true'),'activation was not distinguished from existence')
        r.units['test-1'].air=true; r.event(3,r.units['test-1'])
        r.advance(1)
        assert(r.contains('in_air=true') and r.contains('phase=event id=3'))
        r.units['test-1'].alive=false; r.event(8,r.units['test-1'])
        r.event(5,r.units['test-2'])
        assert(r.contains('phase=event id=8') and r.contains('phase=event id=5'))
    end
    r.advance(120)
    for _,elapsed in ipairs({0,1,5,30,120}) do assert(r.contains('phase=sample elapsed='..elapsed..' ')) end
    assert(next(r.handlers)==nil,'observer leaked after final sample')
    assert(r.calls()==1 and #r.state.spawn_queue==0,'diagnostics retried a spawn')
    assert(r.contains('phase=observation_complete'))
end

local delayed=runtime('empty',true)
delayed.spawn(); delayed.materialise('test-1'); delayed.advance(1)
assert(delayed.contains('unit=test-1 present=true'))
delayed.units['test-1']=nil; delayed.advance(5)
assert(delayed.contains('unit=test-1 present=false'))
delayed.e.Unit.getByName=function() error('lookup failed') end
delayed.advance(30)
assert(delayed.contains('lookup=QUERY_ERROR('))
delayed.advance(120)
assert(next(delayed.handlers)==nil)

local broken=runtime('full',true)
broken.e.Airbase.getByName=function() error('airbase unavailable') end
broken.state.base_pos.FARP={x=100,z=200}
broken.spawn()
assert(broken.contains('base=FARP lookup=QUERY_ERROR(') and broken.calls()==1)
broken.units['test-1'].getLife=function() error('stale handle') end
broken.units['test-1'].isActive=function() error('activation query failed') end
broken.advance(1)
assert(broken.contains('life=QUERY_ERROR(') and broken.contains('active=QUERY_ERROR(') and broken.contains('unit=test-2 present=true'))
broken.e._DMT_GEN=broken.e._DMT_GEN+1
local log_count=#broken.logs
broken.event(3,broken.units['test-1']); broken.advance(120)
assert(#broken.logs==log_count and next(broken.handlers)==nil,'stale generation emitted diagnostics')

local quiet=runtime('full',false)
quiet.spawn()
assert(#quiet.timers==0 and next(quiet.handlers)==nil and not quiet.contains('[spawn_probe]'))
local disabled=runtime('full',true)
disabled.spawn(); disabled.e._DMT_DEBUG=false; disabled.advance(120)
assert(next(disabled.handlers)==nil)

local inactive=runtime('full',true)
inactive.spawn()
inactive.units['test-1'].active=false
inactive.advance(1)
assert(inactive.contains('unit=test-1 present=true exists=true active=false'), 'inactive unit was not distinguished from absence or a query error')

local parking=runtime('full',true)
parking.state.base_pos.FARP={x=100,z=200}
parking.e.Airbase.getByName=function()
    return {
        getID=function() return 77 end,
        getDesc=function() return {category=1} end,
        getCoalition=function() return 1 end,
        getPosition=function() return {p={x=100,y=20,z=200}} end,
        getParking=function(_,available)
            if available then error('free parking unsupported') end
            return {{Term_Index=1,Term_Type=40,vTerminalPos={x=101,y=20,z=201},TO_AC=false}}
        end,
    }
end
parking.e.StaticObject.getByName=function() return {getID=function() return 99 end} end
parking.spawn()
assert(parking.contains('base=FARP id=77 category=1 coalition=1 static_id=99'))
assert(parking.contains('parking=1:40:101,20,201:TO_AC=false'))
assert(parking.contains('parking_available_only=true parking=QUERY_ERROR('))
assert(parking.calls()==1,'parking query failure prevented submission')

local observer_failure=runtime('full',true)
observer_failure.e.world.addEventHandler=function() error('registration failed') end
observer_failure.spawn(); observer_failure.advance(120)
assert(observer_failure.calls()==1 and observer_failure.contains('phase=diagnostic_error'))
assert(observer_failure.contains('phase=sample elapsed=120'))

local scheduling_failure=runtime('full',true)
scheduling_failure.e.timer.scheduleFunction=function() error('scheduling failed') end
scheduling_failure.spawn()
assert(scheduling_failure.calls()==1 and next(scheduling_failure.handlers)==nil)

-- The shared FARP adapter preserves explicit cold and hot choices for both types.
local r=runtime('full',false)
local adapter=r.require_module('farp_parking')
r.state.base_kind.FARP='farp'
local base={getID=function() return 77 end,getDesc=function() return {category=1} end,getPosition=function() return {p={x=100,y=20,z=200}} end}
for _,aircraft in ipairs({'Mi-24V','AH-64D'}) do
    for _,hot in ipairs({false,true}) do
        local p={x=0,y=0,type=hot and 'TakeOffParkingHot' or 'TakeOffParking',action='stale',airdromeId=77}
        local data={name='adapter',units={{name='a',type=aircraft,x=0,y=0}}}
        adapter.configureDeparture('FARP',base,data,p)
        assert(p.type==(hot and 'TakeOffParkingHot' or 'TakeOffParking'))
        assert(p.action==(hot and 'From Parking Area Hot' or 'From Parking Area'))
        assert(p.helipadId==77 and p.linkUnit==77 and p.airdromeId==nil)
    end
end
print('PASS spawn diagnostics: API failures, empty/partial/delayed/disappearing units, events, query failures, sampling, cleanup, generations, debug gating and hot/cold FARP departures')

File: packages/ee-mission/src/spawn_diagnostics.ts
/** @noSelfInFile */
// DCS engine-boundary diagnostics, not EECH campaign policy. Observation times
// come from issue #3's reproduction plan; they never trigger retries or refunds.
import * as cs from "./campaign_state";
import type { SpawnItem } from "./campaign_types";

type GroupSpawn = Extract<SpawnItem, { kind: "group" }>;
type Logger = (this: void, message: string) => void;
interface Probe {
	item: GroupSpawn;
	started: number;
	log: Logger;
}
const SAMPLE_SECONDS = [1, 5, 30, 120];
// Ephemeral observers belong to this loaded module, not persisted campaign state.
const watched = new Map<string, Probe>();
let listening = false;

function emit(probe: Probe, message: string): void {
	pcall(
		probe.log,
		`[spawn_probe] gen=${cs.GENERATION} seq=${probe.item.seq} group=${probe.item.name} ${message}`,
	);
}

function read(query: () => unknown): string {
	const [ok, value] = pcall(query);
	return ok ? tostring(value) : `QUERY_ERROR(${tostring(value)})`;
}

function attributes(unit: Unit): string {
	const result: string[] = [];
	for (const [name, enabled] of pairs(unit.getDesc().attributes)) {
		if (enabled) result.push(name);
	}
	result.sort();
	return result.join(",");
}

function sample(probe: Probe): void {
	emit(
		probe,
		`phase=sample elapsed=${timer.getTime() - probe.started} group_exists=${read(() => Group.getByName(probe.item.name)?.isExist() ?? false)} group_size=${read(() => Group.getByName(probe.item.name)?.getSize())}`,
	);
	for (const expected of probe.item.data.units) {
		const [ok, unit] = pcall(() => Unit.getByName(expected.name));
		if (!ok) {
			emit(
				probe,
				`unit=${expected.name} lookup=QUERY_ERROR(${tostring(unit)})`,
			);
			continue;
		}
		if (!unit) {
			emit(probe, `unit=${expected.name} present=false`);
			continue;
		}
		emit(
			probe,
			`unit=${expected.name} present=true exists=${read(() => unit.isExist())} active=${read(() => unit.isActive())} life=${read(() => unit.getLife())} in_air=${read(() => unit.inAir())} position=${read(
				() => {
					const p = unit.getPoint();
					return `${p.x},${p.y},${p.z}`;
				},
			)} agl=${read(() => {
				const p = unit.getPoint();
				return p.y - land.getHeight({ x: p.x, y: p.z });
			})} attributes=${read(() => attributes(unit))}`,
		);
	}
}

const handler: DcsEventHandler = {
	onEvent(event) {
		if (_DMT_GEN !== cs.GENERATION || !_DMT_DEBUG) return;
		if (
			event.id !== world.event.S_EVENT_BIRTH &&
			event.id !== world.event.S_EVENT_TAKEOFF &&
			event.id !== world.event.S_EVENT_DEAD &&
			event.id !== world.event.S_EVENT_CRASH
		)
			return;
		const [ok, name] = pcall(() => event.initiator?.getName());
		if (!ok || !name) return;
		const probe = watched.get(name);
		if (probe)
			emit(
				probe,
				`phase=event id=${event.id} unit=${name} elapsed=${timer.getTime() - probe.started}`,
			);
	},
};

function describeBase(probe: Probe): void {
	const departure = probe.item.data.route?.points[0];
	const reference = departure?.helipadId ?? departure?.airdromeId;
	if (reference === undefined) {
		emit(probe, "base_reference=none (ground/air start)");
		return;
	}
	let matched = false;
	for (const name of Object.keys(cs.S.base_pos)) {
		const [ok, base] = pcall(() => Airbase.getByName(name));
		if (!ok) {
			emit(probe, `base=${name} lookup=QUERY_ERROR(${tostring(base)})`);
			continue;
		}
		if (!base || read(() => base.getID()) !== tostring(reference)) continue;
		matched = true;
		emit(
			probe,
			`base=${name} id=${reference} category=${read(() => base.getDesc().category)} coalition=${read(() => base.getCoalition())} static_id=${read(() => StaticObject.getByName(name)?.getID())} position=${read(
				() => {
					const p = base.getPosition().p;
					return `${p.x},${p.y},${p.z}`;
				},
			)}`,
		);
		for (const available of [false, true]) {
			emit(
				probe,
				`base=${name} parking_available_only=${available} parking=${read(() =>
					base
						.getParking(available)
						.map(
							(spot) =>
								`${spot.Term_Index}:${spot.Term_Type}:${spot.vTerminalPos.x},${spot.vTerminalPos.y},${spot.vTerminalPos.z}:TO_AC=${spot.TO_AC}`,
						)
						.join(";"),
				)}`,
			);
		}
	}
	if (!matched)
		emit(probe, `base_reference=${reference} matching_campaign_airbase=false`);
}

function forget(probe: Probe): void {
	for (const unit of probe.item.data.units) {
		if (watched.get(unit.name) === probe) watched.delete(unit.name);
	}
	if (watched.size === 0 && listening) {
		pcall(() => world.removeEventHandler(handler));
		listening = false;
	}
}

/** Called before the real API, so synchronous birth events are captured. */
export function begin(item: SpawnItem, log: Logger): Probe | undefined {
	if (
		!_DMT_DEBUG ||
		item.kind !== "group" ||
		item.category !== Group.Category.HELICOPTER
	)
		return undefined;
	const probe = { item, started: timer.getTime(), log };
	const [ok, failure] = pcall(() => {
		for (const unit of item.data.units) watched.set(unit.name, probe);
		if (!listening) {
			world.addEventHandler(handler);
			listening = true;
		}
		const p = item.data.route?.points[0];
		emit(
			probe,
			`probe_version=1 submitted_at=${probe.started} launch_base=${cs.S.group_launch_base[item.name] ?? "unregistered"} deferred_set_task=${item.deferred_setTask !== undefined} deferred_push_task=${item.deferred_pushTask !== undefined}`,
		);
		emit(
			probe,
			`phase=submitted country=${item.country} expected=${item.data.units.length} type=${tostring(p?.type)} action=${tostring(p?.action)} airdrome=${tostring(p?.airdromeId)} helipad=${tostring(p?.helipadId)} link=${tostring(p?.linkUnit)} hidden=${tostring(item.data.hidden)} late_activation=${tostring(item.data.lateActivation)} uncontrolled=${tostring(item.data.uncontrolled)} start_time=${tostring(item.data.start_time)} departure=${tostring(p?.x)},${tostring(p?.y)},${tostring(p?.alt)}`,
		);
		for (const unit of item.data.units) {
			emit(
				probe,
				`requested_unit=${unit.name} type=${unit.type} parking=${tostring(unit.parking)} parking_id=${tostring(unit.parking_id)} position=${unit.x},${unit.y},${tostring(unit.alt)} fuel=${tostring(unit.payload?.fuel)} pylons=${unit.payload?.pylons?.map((pylon) => `${pylon.num}:${pylon.CLSID}`).join(",") ?? "none"}`,
			);
		}
		describeBase(probe);
	});
	if (!ok) emit(probe, `phase=diagnostic_error detail=${tostring(failure)}`);
	return probe;
}

export function finish(
	probe: Probe | undefined,
	outcome: "returned" | "nil" | "threw",
): void {
	if (!probe) return;
	const [ok, failure] = pcall(() => {
		emit(
			probe,
			`phase=api_result outcome=${outcome} (not proof of materialisation)`,
		);
		sample(probe);
		let index = 0;
		timer.scheduleFunction(
			() => {
				if (_DMT_GEN !== cs.GENERATION || !_DMT_DEBUG) {
					forget(probe);
					return undefined;
				}
				const [sampled, reason] = pcall(() => sample(probe));
				if (!sampled)
					emit(probe, `phase=diagnostic_error detail=${tostring(reason)}`);
				index++;
				if (index === SAMPLE_SECONDS.length) {
					emit(
						probe,
						"phase=observation_complete (no failure or refund inferred)",
					);
					forget(probe);
					return undefined;
				}
				return probe.started + SAMPLE_SECONDS[index];
			},
			undefined,
			probe.started + SAMPLE_SECONDS[0],
		);
	});
	if (!ok) {
		emit(probe, `phase=diagnostic_error detail=${tostring(failure)}`);
		forget(probe);
	}
}
